node-repository/src/test/java/com/yahoo/vespa/hosted/provision/maintenance/HostDeprovisionMaintainerTest.java

```diff
@@ -0,0 +1,98 @@
+        tester.addNode("host4-1", Optional.of("host4"), NodeType.tenant, Node.State.reserved, Optional.of(tenantApp));
+
+        Node host2 = tester.addNode("host2", Optional.empty(), NodeType.host, Node.State.failed, Optional.of(tenantApp));
+        tester.addNode("host2-1", Optional.of("host2"), NodeType.tenant, Node.State.failed, Optional.empty());
+
+        assertEquals(7, tester.nodeRepository().getNodes().size());
+
+        maintainer.maintain();
+
+        assertEquals(5, tester.nodeRepository().getNodes().size());
+        verify(hostProvisioner).deprovision(eq(host2));
+        verifyNoMoreInteractions(hostProvisioner);
+
+        assertTrue(tester.nodeRepository().getNode("host2").isEmpty());
+        assertTrue(tester.nodeRepository().getNode("host2-1").isEmpty());
+    }
+
+    @Test
+    public void does_not_remove_if_failed() {
+        Node host2 = tester.addNode("host2", Optional.empty(), NodeType.host, Node.State.failed, Optional.of(tenantApp));
+        doThrow(new RuntimeException()).when(hostProvisioner).deprovision(eq(host2));
+
+        maintainer.maintain();
+
+        assertEquals(1, tester.nodeRepository().getNodes().size());
+        verify(hostProvisioner).deprovision(eq(host2));
+        verifyNoMoreInteractions(hostProvisioner);
+    }
+
+    @Test
+    public void finds_nodes_that_need_deprovisioning() {
+        Node host2 = createNode("host2", Optional.empty(), NodeType.host, Node.State.failed, Optional.of(tenantApp));
+        Node host21 = createNode("host2-1", Optional.of("host2"), NodeType.tenant, Node.State.failed, Optional.empty());
+        Node host3 = createNode("host3", Optional.empty(), NodeType.host, Node.State.provisioned, Optional.empty());
+
+        List<Node> nodes = List.of(
+                createNode("host1", Optional.empty(), NodeType.host, Node.State.active, Optional.of(tenantHostApp)),
+                createNode("host1-1", Optional.of("host1"), NodeType.tenant, Node.State.reserved, Optional.of(tenantApp)),
+                createNode("host1-2", Optional.of("host1"), NodeType.tenant, Node.State.failed, Optional.empty()),
+
+                host2, host21, host3,
+
+                createNode("host4", Optional.empty(), NodeType.host, Node.State.provisioned, Optional.empty()),
+                createNode("host4-1", Optional.of("host4"), NodeType.tenant, Node.State.reserved, Optional.of(tenantApp)),
+
+                createNode("proxyhost1", Optional.empty(), NodeType.proxyhost, Node.State.provisioned, Optional.empty()),
+
+                createNode("proxyhost2", Optional.empty(), NodeType.proxyhost, Node.State.active, Optional.of(proxyHostApp)),
+                createNode("proxy2", Optional.of("proxyhost2"), NodeType.proxy, Node.State.active, Optional.of(proxyApp)));
+
+        Set<Node> expected = Set.of(host2, host3);
+        Set<Node> actual = HostDeprovisionMaintainer.actionableNodes(new NodeList(nodes));
```

Comment:
Why not test using `hostProvisioner.maintain()` directly?

Reply:
The purpose of this test is to verify that given a set of all nodes, the correct ones are picked for processing. It is much easier (don't have to set up mock) and more correct (verifying the method directly, rather than observing which external methods were invoked) to call this directly.